非同期処理の実装

junichiohta opened the pull request:
- callback.js: コールバック方式
- promise.js: Promise方式
- async-await.js: async/await方式
- db-helpers.js: Promise化ヘルパー関数

junichiohta committed:
非同期処理の実装

- callback.js: コールバック方式
- promise.js: Promise方式
- async-await.js: async/await方式
- db-helpers.js: Promise化ヘルパー関数

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env node
+
+import sqlite3 from "sqlite3";
+const { Database } = sqlite3.verbose();
```
import と本文の間は空行を入れてください。

cafedomancer replied:
destruct して `Database` を取り出す必要はないような。プログラム中で `Database` とだけ見ても node-sqlite3 のものであることが分かりづらくなっています。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+  );
+}
+
+runWithoutErrors();
```
トップレベルの処理と関数定義を交互に混在させないでください。読みづらいです。関数定義をまとめて行い、その後にトップレベルの処理を書いてください。また、関数がその外部の変数 (`db`) に依存するのは良くないです。関数の独立性・再利用性が下がるからです。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+      id INTEGER PRIMARY KEY AUTOINCREMENT,
+      title TEXT NOT NULL UNIQUE
+    )
+  `,
```
この表記だと先頭や行頭、末尾に余分な空行や空白が追加され、以下のような文字列が生成されてしまいます。実行上は問題なくても無駄のない文字列を定義してください。

```

    CREATE TABLE books (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      title TEXT NOT NULL UNIQUE
    )
  
```

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+        console.error("テーブル作成エラー:", err.message);
+        db.close();
+        return;
+      }
```
エラーなしのプログラムのエラーハンドリングは不要です。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+      }
+
+      db.run(
+        `INSERT INTO books (title) VALUES (?)`,
```
文字列中に変数を埋め込んでいるわけではないのにテンプレート文字列が使われています。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+            db.close();
+            return;
+          }
+          const insertedId = this.lastID;
```
INSERT された結果として ID を出力する処理がないです。

cafedomancer replied:
`this.lastID` をそのまま参照すればよいだけなので変数に入れる必要はないような。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+                  return;
+                }
+
+                runWithErrors();
```
実行順が分かりづらくなるので Docs にあるように `setTimeout()` を使って実行順を制御してください。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+            console.error("データ挿入エラー:", err.message);
+            db.close();
+            return;
+          }
```
エラーを起こさない想定の部分のエラーハンドリングは不要です。

cafedomancer commented on 03.asynchronous/callback.js:
```diff
@@ -0,0 +1,120 @@
+
+              db.get(`SELECT * FROM books WHERE id = ?`, [999], (err, book) => {
+                if (err) {
+                  console.error("データ取得エラー:", err.message);
```
`指定されたレコードが見つかりません` と表示されているのでエラーが起きてないです。エラーを起こしてください。

junichiohta committed:
レビュー指摘の修正

callback.jsのレビュー指摘に対応。
同様の改善をpromise.js/async-await.jsにも適用。

【callback.jsの指摘事項】
- import後に空行を追加
- sqlite3.Databaseを直接使用
- 関数定義と実行を分離
- 関数が外部変数に依存しないよう引数で受け取る
- SQL文の余分な空白を削除
- エラーなしプログラムのエラーハンドリングを削除
- INSERT結果のIDを出力
- 変数を使わずthis.lastIDを直接使用
- setTimeout()で実行順を制御
- nonexistent_tableで実際のエラーを発生

【追加対応】
- package.jsonにsqlite3依存関係を追加
